### PR TITLE
Refactor wallet autologin

### LINKED ISSUE #35: Refactor Wallet Autologin
Auto login current takes a good couple of seconds, because of the way web3-onboard works, meaning all user-based queries have to wait until we can determine the wallet is connected, which takes up precious loading time

So speed things up, if we detect we're able to auto-login the user, we should do so instantly, without waiting for web3-connect to complete (with error handling of course, in case things go wrong) so that we may be able to fire off queries / mutations before it's done loading in the background

### PR DESCRIPTION
**Pull request by willm30**
## Description

This PR implements a "quick login" functionality to the wallet connection process. If there is wallet information saved in local storage when setting up the user context (e.g. after a logged-in user refreshes the browser), instead of waiting for a full login, we store a simplified wallet in context with address and chain information. This allows queries to run and the app to be interacted with.

Then, we run the standard login via web3-onboard in the background, and once that completes store a full wallet in context.  

# Testing
- Log in and refresh the browser. Wallet connection should be instant and the app fully interactive right away.
- Log in via metamask -> same story.
- Log in via metamask to an unsupported chain. `useCanInteractWithNetwork` should return false, which should restrict what you can do (e.g. can't create a colony). 
- If you change `lastWallet` on line 22 of `setupWalletContext` to a `LastWallet` with a random `type` key, you will be able to test what happens when the wallet stored in localstorage isn't recognised (for whatever reason). You'll be prompted to select a wallet from the ui and if you don't, you'll be logged out. 

**Changes** 🏗

* Added "quick login" functionality to `setupUserContext`
* Added an effect to the `Wallet` which watches for wallet changes and initiates the full login after the quick login is complete

TODO: 
- [x] error handling
- [x] case where last wallet saved is not recognized, hence the autologin logic will fail
- [x] improve local storage for the last wallet type
- [x] Call `onboard.disconnect()` when signing out
~~- [ ] remove onboard.js:agreement key -> See comment in code at `wallet/onboard`~~
- [x] Handle case where user switches to unsupported chain #183

Resolves #35

### CODE REVIEW

**Comment by willm30** [comment]
@rdig I've been experimenting with this, and I have a solution that seems to satisfy the main requirement of #35  (speed things up). 

I just want to check this is what you had in mind when you say:

> So speed things up, if we detect we're able to auto-login the user, we should do so instantly, without waiting for web3-connect to complete 

I'm setting a simplified wallet into context that includes address and chain information. Chain info is required for `useCanInteractWithNetwork` and address for all the user queries. As far as I can tell, these are the only two pieces of information needed (at least at this iteration of the CDapp) for the user to be able to interact with the UI, thus creating an "instant auto-login" effect. 

Before I finish up the pr, I want to confirm: am I on the right lines with this "quick login" idea?

File: src/components/frame/LandingPage/LandingPage.tsx
Comment (willm30) [inline src/components/frame/LandingPage/LandingPage.tsx:70]: Prevents the subscribed colonies list flickering when the full login completes

File: src/redux/sagas/wallet/index.ts
Comment (rdig) [reply src/redux/sagas/wallet/index.ts:124]: No no. This logic doesn't exist anymore.

A user can connect to whatever chain they want. But they can only interact with colonies that were deployed on the same chain as their currently connected wallet.

File: src/redux/sagas/wallet/onboard.ts
Comment (rdig) [reply src/redux/sagas/wallet/onboard.ts:57]: I never bothered with their agreement settings as I really didn't care for them.

But I'm really glad you did :)

File: src/utils/autoLogin.ts
Comment (rdig) [reply src/utils/autoLogin.ts:21]: Very nice. I like this improvement

File: src/utils/autoLogin.ts
Comment (rdig) [reply src/utils/autoLogin.ts:26]: It was never needed, also, as it turns out, I screwed up the logic check :)

**Review by rdig — CHANGES_REQUESTED** [review]
This is excellent work :100: 

There's two reasons I rejected this:
- we don't need to provide the rpc url anymore, the user will
- user can connect to any network with their wallet, doesn't matter if we support it or not

Let me know if something is not clear and maybe we'll have a call detailing it out.

File: src/redux/sagas/wallet/index.ts
Comment (willm30) [reply src/redux/sagas/wallet/index.ts]: Removed

File: src/context/AppContext.tsx
Comment (willm30) [inline src/context/AppContext.tsx:66]: Gets the checksummed address, which is how we store the user on user creation. Noticed that when creating a user when logged in via metamask, the address provided was all lowercase, and searching in the db is case sensitive.

**Review by willm30 — COMMENTED** [review]
@rdig I have removed the chain switching logic and reverted the onboard config back to only include ganache. Please re-review when you get a chance 👍

File: src/redux/sagas/wallet/onboard.ts
Comment (willm30) [inline src/redux/sagas/wallet/onboard.ts:46]: The onboard config only uses the `chains` array to instantiate non-injected wallet modules (like `ganacheModule`), and for programmatic chain switching. We're not doing any programmatic chain switching, and we only have one non-injected module (`ganacheModule), so we only need to include one chain (ganache) in the config.

**Review by jakubcolony — COMMENTED** [review]
Great improvement! All looks good, just wanted to ask you about a specific scenario:
I signed out. The `autologin` local storage item was cleared, but I copied it earlier and pasted a modified version back to storage:
```
{"type":"Dev Wallet 1","address":"0xb77D57F4959eAfA03241394df24b83FcFaf9c15407461xxxxx"}
```
I modified the address by adding a few characters at the end (making it an invalid one). When I refreshed the page, I saw the modified address for a brief period, then it switched to the Dev Wallet 1 without any UI prompt. Is that the expected behaviour? I would expect to see the wallet selection UI in this case.

https://user-images.githubusercontent.com/112586815/214152269-ae243f1d-0828-417f-8b79-cb2d0439c30e.mov

**Comment by willm30** [comment]
@jakubcolony Hmm, I think it only makes sense to speak of expected behaviour in the context of expected user behaviour. I don't think manually editing local storage is an expected user behaviour, and so don't think it's worth making contingencies for such edge cases.

File: src/context/AppContext.tsx
Comment (willm30) [reply src/context/AppContext.tsx:66]: That's what I'm going for. Where would you put it?

File: src/types/definitions/index.d.ts
Comment (willm30) [reply src/types/definitions/index.d.ts:2]: Thanks

**Comment by willm30** [comment]
But from a cursory inspection, the invalid wallet address is causing the user query to fail in AppContext and I presume that's interfering with the login flow somehow (meaning we don't actually arrive at the logic to open the connect wallet modal). 

I can't see this being a problem though.

**Comment by jakubcolony** [comment]
It might be dev only, but it looks surprising that the default method of handling malformatted data is to sign in with the first available wallet, rather than signing out/not logging at all

**Comment by willm30** [comment]
@jakubcolony I do agree with you in the sense that it should be able to handle unexpected behaviour gracefully / logically. I just wonder if it's worth investigating in this case, because it seems unlikely metamask will be returning malformed wallet addresses. 

With that in mind, is this stopping you from approving? If so, I can take a deeper look.

**Review by jakubcolony — APPROVED** [review]
@willm30 
>because it seems unlikely metamask will be returning malformed wallet addresses

Can you be sure though? All I think we need is to gracefully handle those by doing nothing, rather than some unexpected action, especially when it comes to authentication

Whether it's worth investigating, I don't think I can tell. This looks like dev-only issue. I'm happy to approve, and Raul can probably help with a second opinion 👍

**Comment by Julianahlk** [comment]
@willm30  Was trying to test this but was getting an error using the temp-create-data script? TypeError: Cannot destructure property 'abi' of 'colonyJSExtras.factories.latest.MetaTxToken__factory' as it is undefined.

Maybe this needs to be rebased from master first?

**Comment by willm30** [comment]
@rdig I have rebased this onto the QA branch and it seems to work fine. It also appears to solve a couple of problems, e.g the modal appearing when creating a new user and the button sticking if there's already data in local storage. 

Can you sense check to make sure I've not inadvertantly overwritten anything you'd done on the main branch? 

The only material difference I can see is the setting of the chain array in wallet/onboard.ts, but if my comment in that file still stands, we don't need to manually specify the details for Gnosis chain.

File: src/redux/sagas/wallet/onboard.ts
Comment (rdig) [reply src/redux/sagas/wallet/onboard.ts:59]: Are you sure? I couldn't get QA running on Gnosis without providing a chain. Granted I didn't try to just give it the local ganache chain details.

**Review by rdig — APPROVED** [review]
All good from me. Let me know if you want me to pull the changes and push it to QA

**Review by rdig — APPROVED** [review]
All good from me. Let me know if you want me to pull the changes and push it to QA

**Review by rdig — APPROVED** [review]
All good from me. Let me know if you want me to pull the changes and push it to QA

**Review by rdig — APPROVED** [review]
All good from me. Let me know if you want me to pull the changes and push it to QA

**Comment by willm30** [comment]
@rdig Thanks, I will ping you later today as I'm accumulating fixes here: https://github.com/JoinColony/colonyCDapp/pull/533

**Pull request merged**
